src/ol/layer/group.js

```diff
@@ -221,6 +221,9 @@ ol.layer.Group.prototype.getLayerStatesArray = function(opt_states) {
         layerState.extent = ownLayerState.extent;
       }
     }
+    if (layerState.zIndex === 0 && ownLayerState.zIndex !== 0) {
```

Comment:
@Patchez, the user may set a `zIndex` of 0 so it cannot be used to mean `no value`. Instead, the zIndex should be tested against `undefined`.

You will have to change the `zIndex` initialization at https://github.com/openlayers/openlayers/blob/master/src/ol/layer/base.js#L36 and change places where the zIndex is used to fallback to 0 if `undefined`.